docs/address_lookup/the_address_lookup_process.md

```diff
@@ -0,0 +1,95 @@
+# The address lookup process
+
+The address lookup process consists of 4 main parts:
+
+  1. The user enters a postcode. The Ordnance Survey (OS) Places API is then called and a number of addresses with that postcode are returned.  These are then displayed to the user in a dropdown.
+
+  2. The user selects an address from the dropdown. This captures the entire address object, including the Unique Property Reference Number (UPRN). The address is then converted from the format as presented by the API into the required one.
+
+  3. The user can edit the address.  It is desirable to maintain the UPRN irrespective of the changes made by the user. So, it is removed only when it is absolutely necessary to do so, i.e. we can no longer trust that the address as entered still matches the address associated with that UPRN.
+
+  4. The user submits the address to be stored.
+
+There are two points in the process when the user can choose to opt-out of selecting an address, by clicking a link and then be presented with a form into which they can enter their address manually.  This link is shown:
+
+  1. if no results are returned for the entered postcode, and
+
+  2. at the same time as the dropdown, so that the user can click it in cases where the desired address is not present in the dropdown.
+
+In cases where an address is manually entered, the UPRN is not captured, so will be empty in the final stored data.
+
+
+## Conversion
+
+The OS Places API returns addresses in the following structure:
+
+| OS Places API Structure |
+|-------------------------|
+| UPRN                    |
+| SAO_START_NUMBER        |
+| SAO_START_SUFFIX        |
+| SAO_END_NUMBER          |
+| SAO_END_SUFFIX          |
+| SAO_TEXT                |
+| PAO_START_NUMBER        |
+| PAO_START_SUFFIX        |
+| PAO_END_NUMBER          |
+| PAO_END_SUFFIX          |
+| PAO_TEXT                |
+| ORGANISATION            |
+| STREET_DESCRIPTION      |
+| LOCALITY_NAME           |
+| TOWN_NAME               |
+| ADMINISTRATIVE_AREA     |
+| AREA_NAME               |
+| POSTCODE_LOCATOR        |
+
+Stored addresses are required to be in the this structure:
+
+| Required Structure         |
+|----------------------------|
+| uprn                       |
+| building_and_street_line_1 |
+| building_and_street_line_two |
+| town_city                  |
+| county                     |
+| postcode                   |
+
+As the two structures differ significantly, conversion is required between the two.  The following diagram describes the conversion logic.
+
+![Address conversion](images/address_conversion.png)
+
+<p><center><strong>Address conversion</strong></center></p>
+
+## Primary and Secondary address objects
+
+The OS Places API contains two blocks of sub-address elements: one set beginning `PAO_` and the other set beginning `SAO_`.  These identify the Primary Address Object (PAO) and the Secondary Address Object (SAO). All of these fields are optional for an address.
```

Comment:
Does all of them being optional mean that there can exist an address without any PAO or SAO?

Reply:
Yes, in theory it's possible to get an address without any PAO_ or SAO_ fields.  In practice, I'm not sure it'll happen and I've not found one yet.